lib/Service/BotService.php

```diff
@@ -0,0 +1,285 @@
+				'mediaType' => 'text/markdown', // FIXME or text/plain when markdown is disabled
+			],
+			'target' => [
+				'type' => 'Collection',
+				'id' => $event->getRoom()->getToken(),
+				'name' => $event->getRoom()->getName(),
+			]
+		]);
+	}
+
+	public function afterSystemMessageSent(ChatEvent $event, MessageParser $messageParser): void {
+		$bots = $this->getBotsForToken($event->getRoom()->getToken());
+		if (empty($bots)) {
+			return;
+		}
+
+		$message = $messageParser->createMessage(
+			$event->getRoom(),
+			null,
+			$event->getComment(),
+			$this->l10nFactory->get('spreed', 'en', 'en')
+		);
+		$messageParser->parseMessage($message);
+		$messageData = [
+			'message' => $message->getMessage(),
+			'parameters' => $message->getMessageParameters(),
+		];
+
+		$this->sendAsyncRequests($bots, [
+			'type' => 'Activity',
+			'actor' => [
+				'type' => 'Person',
+				'id' => $message->getActorType() . '/' . $message->getActorId(),
+				'name' => $message->getActorDisplayName(),
+			],
+			'object' => [
+				'type' => 'Note',
+				'id' => $event->getComment()->getId(),
+				'name' => $message->getMessageRaw(),
+				'content' => json_encode($messageData),
+				'mediaType' => 'text/markdown',
+			],
+			'target' => [
+				'type' => 'Collection',
+				'id' => $event->getRoom()->getToken(),
+				'name' => $event->getRoom()->getName(),
+			]
+		]);
+	}
+
+	/**
+	 * @param Bot[] $bots
+	 * @param array $body
+	 */
+	protected function sendAsyncRequests(array $bots, array $body): void {
+		$jsonBody = json_encode($body, JSON_THROW_ON_ERROR);
+
+		foreach ($bots as $bot) {
+			$botServer = $bot->getBotServer();
+			$random = $this->secureRandom->generate(64);
+			$hash = hash_hmac('sha256', $random . $jsonBody, $botServer->getSecret());
+			$headers = [
+				'Content-Type' => 'application/json',
+				'X-Nextcloud-Talk-Random' => $random,
+				'X-Nextcloud-Talk-Signature' => $hash,
+				'X-Nextcloud-Talk-Backend' => rtrim($this->serverConfig->getSystemValueString('overwrite.cli.url'), '/') . '/',
+				'OCS-APIRequest' => 'true', // FIXME optional?
+			];
+
+			$data = [
+				'verify' => false,
+				'nextcloud' => [
+					'allow_local_address' => true, // FIXME don't enforce
+				],
+				'headers' => $headers,
+				'timeout' => 5,
+				'body' => json_encode($body),
+			];
+
+			$client = $this->clientService->newClient();
+			$promise = $client->postAsync($botServer->getUrl(), $data);
+
+			$promise->then(function (IResponse $response) use ($botServer) {
+				if ($response->getStatusCode() !== Http::STATUS_OK && $response->getStatusCode() !== Http::STATUS_ACCEPTED) {
+					$this->logger->error('Bot responded with unexpected status code (Received: ' . $response->getStatusCode() . '), increasing error count');
+					$botServer->setErrorCount($botServer->getErrorCount() + 1);
+					$botServer->setLastErrorDate($this->timeFactory->now());
+					$botServer->setLastErrorMessage('UnexpectedStatusCode: ' . $response->getStatusCode());
```

Comment:
Maybe we can add a `registerNewError` method?
Something like: 
```php
public function registerNewError(string $message) {
    ++$this->errorCount;
    $this->lastErrorDate = new \DateTimeImmutable();
    $this->lastErrorMessage = $message;
}
```
Maybe a better name can be found…

Reply:
So, it could be without it:
```php
public function registerNewError(string $message, \DateTime $datetime) {
    ++$this->errorCount;
    $this->lastErrorDate = $datetime;
    $this->lastErrorMessage = $message;
}
```
The idea is to wrap everything related to the new error in the same call